File: q76-100/q78.py
# 2020.09.15
# Problem Statement:
# https://leetcode.com/problems/subsets/

class Solution:
    # the same (almost) function in q77
    def combine(self, n, k, answer, nums):
        stack = []
        x = 1
        while True:
            if len(stack) == k:
                answer.append(list(stack))
            if len(stack) == k or x > n:
                if not stack:
                    break
                # was: x = stack.pop() + 1
                # but want to use the index of the first element in stack instead
                # should add 2, since index = stack.pop()-1 in previous question
                x = nums.index(stack.pop()) + 2
            else:
                stack.append(nums[x-1])
                x += 1
                
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.answer = []
        
        for i in range(1, len(nums)+1):
            self.combine(len(nums), i, self.answer, nums)

        # take care of the empty subset    
        self.answer.append([])
        return self.answer

File: Amazon/q78.java
// 2022.05.28
// Problem Statement:
// https://leetcode.com/problems/subsets/

// idea: recursion, include first element or not to all the subsets of nums[1:]
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        if (nums.length==1) {
            List<List<Integer>> ret = new ArrayList<> ();
            List<Integer> l1 = Arrays.asList(nums[0]);
            ret.add(l1);
            ret.add(new ArrayList<Integer> ());
            return ret;
        }
        
        List<List<Integer>> rest = subsets(Arrays.copyOfRange(nums, 1, nums.length));
        List<List<Integer>> answer = new ArrayList<> ();
        for (List<Integer> l : rest) {
            answer.add(l);
            List<Integer> new_list = new ArrayList<Integer> (l);
            new_list.add(0, new Integer (nums[0]));
            answer.add(new_list);
        }
        return answer;
    }
}